### PR TITLE
Notification of spack_packages update for model image rebuild

### LINKED ISSUE #4: Trigger image rebuilds on ACCESS-NRI/spack_packages updates
Thanks @harshula for the suggestion.

We want to trigger our build images to rebuild whenever ACCESS-NRI/spack_packages has a relevant package file update. For example, if libaccessom2.py is updated we want to rebuild all libaccessom2 build images.

For a first pass implementation just a blanket rebuild on any push to spack_packages would be fine. 

Ideally we would check what package files are updated and only trigger the relevant image rebuild. Will need to handle dependencies (e.g. datetime-fortran, etc) somehow in this as well.

See https://github.com/orgs/community/discussions/26323 for an example of how to trigger workflows from an external repository.

### PR DESCRIPTION
**Pull request by CodeGat**
This PR handles the dispatched event from `ACCESS-NRI/spack_packages` that notifies `build-ci` that it has had some kind of update pushed to it and needs specific images to be rebuilt. 

Related PR from spack_packages: ACCESS-NRI/spack_packages#26

Should close #4 !

### CODE REVIEW

File: .github/workflows/notified-of-spack-packages-update.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      packages: ${{ steps.generate.outputs.packages }}
+      compilers: ${{ steps.generate.outputs.compilers }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with: 
+          python-version: '3.10'
+      - name: Generate model matrix
+        id: generate
+        run: python ./util/matrix-generator.py ${{ github.event.inputs.packages }}
```
Comment (Quartzer2) [inline .github/workflows/notified-of-spack-packages-update.yml:36]: In order to verify this python script, should we have a separate test for it?

File: .github/workflows/notified-of-spack-packages-update.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+          dockerfile-directory: containers
+          dockerfile-name: Dockerfile.build
+          build-args: |
+            # TODO: Probably shouldn't hard code base image path
+            "BASE_IMAGE=ghcr.io/access-nri/base-spack-${{ github.event.inputs.spack-packages-version }}:latest"
+            "PACKAGE=${{ matrix.package }}"
+            "COMPILER_NAME=${{ matrix.compiler.name}}"
+            "COMPILER_PACKAGE=${{ matrix.compiler.package}}"
+            "COMPILER_VERSION=${{ matrix.compiler.version}}"
+    permissions:
```
Comment (Quartzer2) [inline .github/workflows/notified-of-spack-packages-update.yml:62]: We are interested in why permissions need to be explicitly defined. Just interested!

File: util/full-dependency-mapping.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+datetime-fortran libaccessom2 cice5 mom5 access-om2
+oasis3-mct libaccessom2 cice5 mom5 access-om2
+libaccessom2 cice5 mom5 access-om2
+cice5 access-om2
+mom5 access-om2
+access-om2 
+access-om3 
+fcm
+gcom
```
Comment (Quartzer2) [inline util/full-dependency-mapping.txt:9]: How are these updated? Again just for interest!

File: util/supported-dependency-mapping.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+datetime-fortran libaccessom2 cice5 mom5
+oasis3-mct libaccessom2 cice5 mom5
+libaccessom2 cice5 mom5
+cice5
+mom5
+access-om3
```
Comment (Quartzer2) [inline util/supported-dependency-mapping.txt:6]: Is this generated from the full dependency mapping? Do we need instructions on what goes in here and how it is updated?

**Review by Quartzer2 — APPROVED** [review]
Again looks good. Just some questions and we are having fun going through and learning the review procedure :)

File: .github/workflows/notified-of-spack-packages-update.yml (hunk 2 above)
Comment (CodeGat) [reply .github/workflows/notified-of-spack-packages-update.yml:62]: I actually don't know :D It was just how the original was done, so I copied it

File: util/full-dependency-mapping.txt (hunk 3 above)
Comment (CodeGat) [reply util/full-dependency-mapping.txt:9]: For now, they aren't updated. I did the dependent graph manually, through investigating the spack environment. It'd have to be updated manually if the dependents change. Issue raised here: https://github.com/ACCESS-NRI/build-ci/issues/21

File: util/supported-dependency-mapping.txt (hunk 4 above)
Comment (CodeGat) [reply util/supported-dependency-mapping.txt:6]: Yes, it is generated from the full dependency mapping. Essentially it was just removing the models that we don't support at the moment (access-om2 among others). I will add a readme to the util folder, though,

File: .github/workflows/notified-of-spack-packages-update.yml (hunk 1 above)
Comment (CodeGat) [reply .github/workflows/notified-of-spack-packages-update.yml:36]: Yes, we should. I'll add one.

**Review by Quartzer2 — APPROVED** [review]

**Pull request merged**